Support attribute read/write in Android

austinh0 opened the pull request:
#### Problem
* No attribute support in Java generated clusters code (#7376)

#### Change overview
* Add template code generating `read`/`write` methods, `*AttributeCallback` interfaces, and `*Attribute` classes for struct response types. 
* In `JniTypeWrappers.h`, add `JniClass` which uses RAII to delete global Java references.
* In Android CHIPTool, add button to read on/off attribute.

#### Testing
Went through commissioning and manually tested read/write methods.

bzbarsky-apple commented on src/app/zap-templates/common/ChipTypesHelper.js:
```diff
@@ -36,6 +36,32 @@ function asBasicType(type)
   }
 }
 
+function asCallbackAttributeBasicType(type)
```
@vivien-apple could you take a look please?

austinh0 commented on src/controller/java/JniTypeWrappers.h:
```diff
@@ -58,6 +62,14 @@ class UtfString
     UtfString(JNIEnv * env, const char * data) : mEnv(env) { mData = data != nullptr ? mEnv->NewStringUTF(data) : nullptr; }
+    UtfString(JNIEnv * env, chip::ByteSpan data) : mEnv(env)
+    {
+        char dataBuf[data.size() + 1];
```
Using ostringstream now.

austinh0 commented:
@Damian-Nordic @vivien-apple @saurabhst @mspang

woody-apple commented:
@mspang @Damian-Nordic @LuDuda @jmartinez-silabs ?

woody-apple commented:
@austinh0 Conflicts here :(

mspang commented on src/controller/java/templates/helper.js:
```diff
@@ -128,6 +146,94 @@ function asJniSignature(label, type)
+  case 'Int16s':
+    return 'Integer';
+  case 'Int32u':
+  case 'Int64u':
```
Is it not a problem that Int64u doesn't fit into a Java long because Java only has signed types?

austinh0 replied:
`uint64_t` can "fit" into a Java long, as long as we only use methods like [divideUnsigned](https://docs.oracle.com/javase/8/docs/api/java/lang/Integer.html#divideUnsigned-int-int-). However, this is error-prone since in the current implementation there's no way to tell whether a signed or unsigned long is returned solely by looking at the code.

I looked into replacing `long` with `BigInteger`, but it seems like overkill to me, and I ran into some problems with the ZAP template using a promise in `{{#if}}` helpers.

I think a good middle ground here could be to create a type annotation to declare that a long field is unsigned. [Checker framework](https://checkerframework.org/manual/#signedness-checker) does something similar with [@Unsigned](https://checkerframework.org/api/org/checkerframework/checker/signedness/qual/Unsigned.html). We wouldn't have the compile-time checking that Checker has, unless we want to force clients to depend on Checker, but at least the signedness would be clear. WDYT?

austinh0 replied:
After some more thinking, this is probably out of scope for the PR, especially for the `BigInteger` solution, which would require a lot of changes in the template. I filed #8074 and added a TODO to track.

vivien-apple commented on src/app/zap-templates/templates/chip/helper.js:
```diff
@@ -271,6 +271,19 @@ function isManufacturerSpecificCommand()
+    }
+  });
+  return asBlocks.call(this, Promise.resolve(validTypes), options);
+}
```
It makes sense to have an iterator for the base callbacks types.

That would be useful for the code gen of other platforms as well.

But instead of making it that way and add more types conversions, Can you add an iterator that consumes all the atomics and create a list of "callbacks" from them ?

That would replace`valid_callback_attribute_types` and `asCallbackAttributeBasicType` by a new iterator that  carries both the original {{type}} and the {{chipType}}. That should allow to remove `attributeCallbackTypeStringAsCType`  and prevent custom checks for types such as `{{#if (isStrEqual type "String")}}` instead of the helpers from `StringHelper.js`

austinh0 commented:
As discussed offline, @vivien-apple  may do some refactoring in the ZAP helpers next week that this PR can benefit from. Will hold until then.

austinh0 commented:
Waiting for #8117

austinh0 commented:
After rebasing on top of #8117:
* Replaced my attribute callback types helper with {{chip_server_global_response}}
* Now reusing conversion helpers from `java/templates/helper.js` instead of defining custom helpers for attributes
* Split `StringAttributeCallback` into `OctetStringAttributeCallback` and `CharStringAttributeCallback`

andy31415 commented:
/rebase
